framework/src/main/java/org/checkerframework/framework/type/QualifierHierarchy.java

```diff
@@ -81,361 +100,351 @@ public int getWidth() {
-
+        assertSameSize(qualifiers1, qualifiers2, result);
         return result;
     }
 
     /**
-     * Tests whether {@code subAnno} is a sub-qualifier of, or equal to, {@code superAnno},
-     * according to the type qualifier hierarchy. This checks only the qualifiers, not the Java
-     * type.
-     *
-     * <p>This method works even if the underlying Java type is a type variable. In that case, a
-     * 'null' AnnnotationMirror and the empty set represent a meaningful value (namely, no
-     * annotation).
+     * Returns true if and only if {@link AnnotatedTypeMirror#getAnnotations()} can return a set
+     * with fewer qualifiers than the width of the QualifierHierarchy.
      *
-     * @param subAnno a qualifier that might be a subtype
-     * @param superAnno a qualifier that might be a subtype
-     * @return true iff {@code subAnno} is a sub qualifier of, or equal to, {@code superAnno}
+     * @param type annotated type mirror
+     * @return true if and only if {@link AnnotatedTypeMirror#getAnnotations()} can return a set
+     *     with fewer qualifiers than the width of the QualifierHierarchy
      */
-    public abstract boolean isSubtypeTypeVariable(
-            AnnotationMirror subAnno, AnnotationMirror superAnno);
+    public static boolean canHaveEmptyAnnotationSet(AnnotatedTypeMirror type) {
+        return type.getKind() == TypeKind.TYPEVAR
+                || type.getKind() == TypeKind.WILDCARD
+                ||
+                // TODO: or should the union/intersection be the LUB of the alternatives?
+                type.getKind() == TypeKind.UNION
+                || type.getKind() == TypeKind.INTERSECTION;
+    }
 
     /**
-     * Tests whether there is any annotation in superAnnos that is a super qualifier of or equal to
-     * some annotation in subAnnos. superAnnos and subAnnos contain only the annotations, not the
-     * Java type.
-     *
-     * <p>This method works even if the underlying Java type is a type variable. In that case, a
-     * 'null' AnnnotationMirror and the empty set represent a meaningful value (namely, no
-     * annotation).
+     * Returns the annotation in annos that is in the same hierarchy as annotationMirror.
      *
-     * @return true iff an annotation in superAnnos is a supertype of, or equal to, one in subAnnos
+     * @param annos set of annotations to search
+     * @param annotationMirror annotation that is in the same hierarchy as the returned annotation
+     * @return annotation in the same hierarchy as annotationMirror, or null if one is not found
      */
-    // This method requires more revision.
-    public abstract boolean isSubtypeTypeVariable(
-            Collection<? extends AnnotationMirror> subAnnos,
-            Collection<? extends AnnotationMirror> superAnnos);
+    default @Nullable AnnotationMirror findAnnotationInSameHierarchy(
+            Collection<? extends AnnotationMirror> annos, AnnotationMirror annotationMirror) {
```

Comment:
Other methods in this interface would call these `qualifiers` and `qualifier`, I think. Let's try to use consistent names within a class/interface. A few other places below.